### ISSUE TITLE
Clarify behavior of a fence after vkQueuePresentKHR() 

### ISSUE THREAD

**Issue by marti4d** [KhronosGroup/Vulkan-Docs#595]
Since `vkQueuePresentKHR()` is mostly a "set-it-and-forget-it" operation from the app side, it doesn't allow setting any semaphores or fences as part of its submission to a queue.

But what guarantees are made about the queued present if I submit a fence to the queue and wait on it, either manually or as a result of `vkWaitQueueIdle()` or `vkWaitDeviceIdle()`?

I **think** that the answer is that the fence would signal once the present request was processed, but before the presentation engine resets the semaphore and presents the image.

I draw my conclusion from this text, but if I'm correct it might be nice for it to be more explicit:

```
The presentation command itself does not delay processing of subsequent commands on the queue, however, presentation requests sent to a particular queue are always performed in order. Exact presentation timing is controlled by the semantics of the presentation engine and native platform in use.
```

**Comment by critsec** [comment]
Right, `vkQueuePresentKHR` followed by `vkWaitQueueIdle` (or equivalent) will return once work submitted to the queue before the `vkQueuePresentKHR` has completed, and any work on the queue needed to present the image (commonly nothing, but in some cases there might be a blit or similar) has completed, and the presentation request has been committed to the presentation engine. That last step is usually pretty lightweight, on the order of a kernel syscall or an asynchronous IPC call.

The language you quote was trying to say this. Do you have a suggestion for what would be more clear?It can't say that the vkQueuePresentKHR command completes instantly (once semaphore waits are satisfied), because that's not always true. It's a bit tricky to say much, since there's a lot of variety in how "presentation engines" behave and how they interact with the GPU, so most things we tried to say that were more precise/descriptive ended up being untrue in some cases.

**Comment by marti4d** [comment]
Thank you for clarifying that - I think perhaps some of what you just said would be nice as a "note" section in the spec.

I think the confusion for me was just around what it means for a present job to be 'completed' from the perspective of the queue it was submitted to. It might be good to explain that - something along the lines of, 

```
For the purposes of queue progress, the `vkQueuePresentKHR()` request is considered complete once any work that the presentation engine *may* require to prepare the image for presentation has been finished, and the presentation request has been offloaded to the presentation engine. 
```

**Comment by cinolt** [comment]
Hello, this paragraph:

> Right, vkQueuePresentKHR followed by vkWaitQueueIdle (or equivalent) will return once work submitted to the queue before the vkQueuePresentKHR has completed, and any work on the queue needed to present the image (commonly nothing, but in some cases there might be a blit or similar) has completed, and the presentation request has been committed to the presentation engine. That last step is usually pretty lightweight, on the order of a kernel syscall or an asynchronous IPC call.

appears to state that vkQueuePresentKHR followed by vkWaitForFences (and thus vkWaitQueueIdle) guarantees that vkQueuePresentKHR has unsignalled its semaphore when vkWaitForFences returns.

I believe this is not implied in the specification, but should be.

When vkWaitForFences waits for a fence, according to the specification, this only implies that vkQueueSubmit's before the "fence signal operation" (and the batches in the same vkQueueSubmit that signalled the fence) completion happen-before vkWaitForFences returns.

Unless I've missed something, vkQueuePresentKHR does not appear to be included in the above dependencies, but should be. Perhaps vkQueuePresentKHR could be defined to be included in some sort of set of operations, that are properly included in the first synchronization scope of a "fence signal operation".
